### PR TITLE
overlord helpers framework and tasklog auto cleanup

### PR DESCRIPTION
**Pull request by himanshug**
We push the task logs to HDFS which fill up the scarce namespace and whose deletion is currently managed outside of Druid. This is a major inconvenience to most of the teams using Druid and most of them are deleting those manually.

This patch introduces a feature to configure Overlord to automatically delete older task logs. See updates to `docs/content/configuration/indexing-service.md` for the user documentation.

Currently, implementation is only provided for local and hdfs type tasklogs.

Framework introduced can be used to manage retention of other tasks related persisted state but that is for the future.

### CODE REVIEW

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelper.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+* under the License.
+*/
+
+package io.druid.indexing.overlord.helpers;
+
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ */
+public interface OverlordHelper
```
Comment (fjy) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelper.java]: While you are in this code, u should also consider merging overlord and coordinator :P

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelperManager.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    synchronized (lock) {
+      if (!started) {
+        log.info("OverlordHelperManager is starting.");
+
+        for (OverlordHelper helper : helpers) {
+          if (helper.isEnabled()) {
+            if (exec == null) {
+              exec = execFactory.create(1, "Overlord-Helper-Manager-Exec--%d");
+            }
+            helper.schedule(exec);
```
Comment (fjy) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelperManager.java]: this logic is basically just repeating what the coordinator is doing

high level comment, can we have this as a coordinator helper?

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelper.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+{
+  boolean isEnabled();
+  void schedule(ScheduledExecutorService exec);
```
Comment (fjy) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelper.java]: i think the method here should be run()

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelper.java (hunk 1 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelper.java]: hmmm, actually we can. I believe we have had the desire for quite some time and it was discussed in the dev sync meeting too and agreed upon.

I would imaging introducing a CliMaster (a "master" node) that binds/start everything that coordinator and overlords do. also, have only one "leader election" for the "leader master" instead of the two.

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelperManager.java (hunk 2 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelperManager.java]: yeah, it is similar. but, i did it on overlord because tasks and tasklogs are owned by overlord and their retention should also be handled at the overlord.

File: docs/content/configuration/indexing-service.md [hunk 4]
```diff
@@ -25,6 +25,15 @@ If you are running the indexing service in remote mode, the task logs must be st
 |--------|-----------|-------|
 |`druid.indexer.logs.type`|Choices:noop, s3, azure, hdfs, file. Where to store task logs|file|
 
+You can also configure the Overlord to automatically retain the task logs only for last x seconds by configuring following additional properties.
+
+|Property|Description|Default|
+|--------|-----------|-------|
+|`druid.indexer.logs.kill.enabled`|Boolean value for whether to enable deletion of old task logs. |false|
+|`druid.indexer.logs.kill.durationToRetain`| Required if kill is enabled. In seconds, task logs to be retained created in last x seconds. |None|
+|`druid.indexer.logs.kill.initialDelay`| Optional. Number of seconds after overlord start when first auto kill is run. |random value less than 300 (5 mins)|
```
Comment (pjain1) [inline docs/content/configuration/indexing-service.md]: random value between 1 and 5 mins ?

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelper.java (hunk 1 above)
Comment (fjy) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelper.java]: @himanshug Yeah that would be pretty great, we don't have to do it as part of this PR though, but I think it would be a really useful to have

File: docs/content/configuration/indexing-service.md (hunk 4 above)
Comment (drcrallen) [inline docs/content/configuration/indexing-service.md]: Same thing about either milliseconds or a Duration

File: docs/content/configuration/indexing-service.md [hunk 5]
```diff
@@ -25,6 +25,15 @@ If you are running the indexing service in remote mode, the task logs must be st
...
+|`druid.indexer.logs.kill.delay`|Optional. Number of seconds of delay between successive executions of auto kill run. |21600 (6 hours)|
```
Comment (drcrallen) [inline docs/content/configuration/indexing-service.md]: Same thing about milliseconds or a string Duration

File: extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/tasklog/HdfsTaskLogs.java [hunk 6]
```diff
@@ -123,6 +125,28 @@ public void killAll() throws IOException
...
   }
+
+  @Override
+  public void kill(long beforeTimestamp) throws IOException
+  {
+    Path taskLogDir = new Path(config.getDirectory());
+    FileSystem fs = taskLogDir.getFileSystem(hadoopConfig);
+    if (fs.exists(taskLogDir)) {
+      RemoteIterator<LocatedFileStatus> iter = fs.listLocatedStatus(taskLogDir);
+      while (iter.hasNext()) {
```
Comment (drcrallen) [inline extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/tasklog/HdfsTaskLogs.java]: There's no failure recovery in this loop, and I'm forgetting how HDFS handles various issues, is this intended to just fail this kill run if there are any sort of issues?

File: extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/tasklog/HdfsTaskLogs.java [hunk 7]
```diff
@@ -123,6 +125,28 @@ public void killAll() throws IOException
...
+        LocatedFileStatus file = iter.next();
+        if (file.getModificationTime() < beforeTimestamp) {
+          Path p = file.getPath();
+          log.info("Deleting hdfs task log [%s].", p.toUri().toString());
+          fs.delete(p, true);
+        }
+
+        if (Thread.interrupted()) {
+          throw new IOException("Thread interrupted. Couldn't delete all tasklogs.");
```
Comment (drcrallen) [inline extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/tasklog/HdfsTaskLogs.java]: How about wrapping an `InterruptedException` in an `IOException`?

File: extensions-core/hdfs-storage/src/test/java/io/druid/indexing/common/tasklogs/HdfsTaskLogsTest.java [hunk 8]
```diff
@@ -68,17 +68,47 @@ public void testOverwrite() throws Exception
...
+    //is necessary to separate 2 file creations by a timestamp that would result in only one
+    //of them getting deleted
+    Thread.sleep(1500);
+    long time = (System.currentTimeMillis()/1000)*1000;
+
+    Files.write("log2content", logFile, Charsets.UTF_8);
+    taskLogs.pushTaskLog("log2", logFile);
+    Assert.assertEquals("log2content", readLog(taskLogs, "log2", 0));
+
+    taskLogs.kill(time);
```
Comment (drcrallen) [inline extensions-core/hdfs-storage/src/test/java/io/druid/indexing/common/tasklogs/HdfsTaskLogsTest.java]: Before here, can you check and make sure the modified time for the old file is before `time` and the new file is after `time`?

**Comment by drcrallen** [comment]
Quasi-related https://github.com/druid-io/druid/issues/401

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelperManager.java (hunk 2 above)
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelperManager.java]: If the executor is having issues and rejecting execution, what is the intended behavior here?

**Comment by drcrallen** [comment]
Also, as a general comment, before the overlord and coordinator are merged, I'd like to see the Helper stuff go into an extensible form.

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelper.java (hunk 1 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelper.java]: yeah, will do that in separate PR. let us keep this one as is.

File: docs/content/configuration/indexing-service.md [hunk 9]
```diff
@@ -25,6 +25,15 @@ If you are running the indexing service in remote mode, the task logs must be st
 |--------|-----------|-------|
 |`druid.indexer.logs.type`|Choices:noop, s3, azure, hdfs, file. Where to store task logs|file|
 
+You can also configure the Overlord to automatically retain the task logs only for last x seconds by configuring following additional properties.
+
+|Property|Description|Default|
+|--------|-----------|-------|
+|`druid.indexer.logs.kill.enabled`|Boolean value for whether to enable deletion of old task logs. |false|
+|`druid.indexer.logs.kill.durationToRetain`| Required if kill is enabled. In seconds, task logs to be retained created in last x seconds. |None|
```
Comment (himanshug) [inline docs/content/configuration/indexing-service.md]: default does not matter because it is forced to be provided by the user when enabled. Long.MAX_VALUE is there just because some value needs to be stored there when it is not enabled.

File: docs/content/configuration/indexing-service.md (hunk 4 above)
Comment (himanshug) [inline docs/content/configuration/indexing-service.md]: dint want to make document too confusing by declaring exact randomness behavior, i think knowing that its some random value less than 5 mins is enough.

**Comment by himanshug** [comment]
@drcrallen 
clock skew between druid nodes and hdfs nodes might end up deleting unintended logs, I will note that in the documentation.

regarding ovelord and coordinator merge, helper stuff introduced here is already in an extensible form.

File: docs/content/configuration/indexing-service.md (hunk 9 above)
Comment (himanshug) [inline docs/content/configuration/indexing-service.md]: changed to milliseconds

File: docs/content/configuration/indexing-service.md (hunk 5 above)
Comment (himanshug) [inline docs/content/configuration/indexing-service.md]: milliseconds

File: extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/tasklog/HdfsTaskLogs.java (hunk 7 above)
Comment (himanshug) [inline extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/tasklog/HdfsTaskLogs.java]: will do

File: api/src/main/java/io/druid/tasklogs/NoopTaskLogs.java [hunk 10]
```diff
@@ -48,4 +48,10 @@ public void killAll() throws IOException
   {
     log.info("Noop: No task logs are deleted.");
   }
+
+  @Override
+  public void kill(long beforeTimestamp) throws IOException
```
Comment (himanshug) [inline api/src/main/java/io/druid/tasklogs/NoopTaskLogs.java]: changed to `void killOlderThan(long timestamp)`

**Comment by himanshug** [comment]
@fjy @drcrallen @pjain1 resolved all comments.

File: extensions-core/hdfs-storage/src/test/java/io/druid/indexing/common/tasklogs/HdfsTaskLogsTest.java (hunk 8 above)
Comment (himanshug) [inline extensions-core/hdfs-storage/src/test/java/io/druid/indexing/common/tasklogs/HdfsTaskLogsTest.java]: done

**Comment by fjy** [comment]
👍

**Comment by himanshug** [comment]
@drcrallen all comments were addressed, pls take a relook

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfig.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+      @JsonProperty("delay") Long delay,
+      @JsonProperty("durationToRetain") Long durationToRetain
+  ){
+    if (enabled) {
+      Preconditions.checkNotNull(durationToRetain, "durationToRetain must be provided.");
+    }
+
+    this.enabled = enabled;
+    this.initialDelay = initialDelay == null ? 60000 + new Random().nextInt(4*60000) : initialDelay.longValue();
+    this.delay = delay == null ? 6*60*60*1000 : delay.longValue();
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfig.java]: Suggest sanity check for positive number.

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfig.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    this.durationToRetain = durationToRetain == null ? Long.MAX_VALUE : durationToRetain.longValue();
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfig.java]: Suggest sanity check for positive number.

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfig.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+      @JsonProperty("initialDelay") Long initialDelay,
+      @JsonProperty("delay") Long delay,
+      @JsonProperty("durationToRetain") Long durationToRetain
+  ){
+    if (enabled) {
+      Preconditions.checkNotNull(durationToRetain, "durationToRetain must be provided.");
+    }
+
+    this.enabled = enabled;
+    this.initialDelay = initialDelay == null ? 60000 + new Random().nextInt(4*60000) : initialDelay.longValue();
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfig.java]: Suggest sanity check for positive number.

File: indexing-service/src/test/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfigTest.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+package io.druid.indexing.overlord.helpers;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.druid.TestUtil;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ */
+public class TaskLogAutoCleanerConfigTest
```
Comment (drcrallen) [inline indexing-service/src/test/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfigTest.java]: Suggest adding test for defaults.

File: indexing-service/src/main/java/io/druid/indexing/common/tasklogs/FileTaskLogs.java [hunk 15]
```diff
@@ -89,4 +90,33 @@ public void killAll() throws IOException
     log.info("Deleting all task logs from local dir [%s].", config.getDirectory().getAbsolutePath());
     FileUtils.deleteDirectory(config.getDirectory());
   }
+
+  @Override
+  public void killOlderThan(final long timestamp) throws IOException
+  {
+    File taskLogDir = config.getDirectory();
+    if (taskLogDir.exists()) {
```
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/common/tasklogs/FileTaskLogs.java]: done

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelperManager.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+ */
+public class OverlordHelperManager
+{
+  private static final Logger log = new Logger(OverlordHelperManager.class);
+
+  private final ScheduledExecutorFactory execFactory;
+  private final Set<OverlordHelper> helpers;
+
+  private volatile ScheduledExecutorService exec;
+  private final Object lock = new Object();
```
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelperManager.java]: done

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfig.java (hunk 13 above)
Comment (himanshug) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfig.java]: done

File: indexing-service/src/test/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfigTest.java (hunk 14 above)
Comment (himanshug) [inline indexing-service/src/test/java/io/druid/indexing/overlord/helpers/TaskLogAutoCleanerConfigTest.java]: done

**Comment by fjy** [comment]
@drcrallen any more comments?

**Comment by fjy** [comment]
👍

File: indexing-service/src/main/java/io/druid/indexing/common/tasklogs/FileTaskLogs.java [hunk 17]
```diff
@@ -89,4 +90,38 @@ public void killAll() throws IOException
...
+
+      if (!taskLogDir.isDirectory()) {
+        throw new IOException(String.format("taskLogDir [%s] must be a directory.", taskLogDir));
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/common/tasklogs/FileTaskLogs.java]: For future reference `IOE` would work best here, not worth blocking on though.

File: indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelperManager.java (hunk 2 above)
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/helpers/OverlordHelperManager.java]: cool
